### PR TITLE
Loop optimization (WIP)

### PR DESCRIPTION
**Pull request by patricksurry**
this is still work in progress but in a consistent state with tests passing if you want to take a peek.

so far it reorganizes zero page a little, making space for loop control (mostly unused so far)
and refactors leave to get it off the return stack.  this simplifies the runtime so we just have simple jmp's 

next task is to route the loop vars thru zp so to simplify the increment and index words.

### CODE REVIEW

File: definitions.asm [hunk 1]
```diff
@@ -46,15 +48,13 @@ cib:        .word buffer0           ; address of current input buffer
 ciblen:     .word 0                 ; length of current input buffer
 toin:       .word 0                 ; pointer to CIB (>IN in Forth)
 
-ip:         .word 0                 ; Instruction Pointer (current xt)
 output:     .word kernel_putc       ; vector for EMIT
 input:      .word kernel_getc       ; vector for KEY
 havekey:    .word 0                 ; vector for KEY?
```
Comment (SamCoVT) [inline definitions.asm:53]: Just a note here - Tali does not currenly implement `KEY?` at all (it expects blocking behavior from `kernel_getc`), so I believe it's totally safe to remove `havekey:` at this point in time.  The issue with KEY? is that not all input sources allow you to check if there is a char available without actually trying to read one, so it was left unimplemented with the assumption that a user could implement it themselves if they wanted it (see #51).

File: taliforth.asm [hunk 2]
```diff
@@ -227,6 +227,20 @@ dovar:
 ; =====================================================================
 ; LOW LEVEL HELPER FUNCTIONS
 
+push_upvar_tos:
```
Comment (SamCoVT) [inline taliforth.asm:230]: I probably should have done this by now.  Just so you are aware, Tali already has a non-standard word `useraddr` which gives the address (in zero page) of `UP`.  Using that, it's possible to pull the base address of the user page varibles, add an offset, and dereference to get access to these values from Forth (does require knowledge of the offsets, of course).

**Comment by SamCoVT** [comment]
I think you've cleaned out enough zp space you could put I _and_ J in zero page.  I gave that some thought, but I don't think it's worth it - you'd have to shuffle J to the return stack and then I to J for each new loop (and then back at the end).

It looks like you are all set up to make the change.  The cycle tests should let you know if it makes sense to go this route (I think it does) and the regular tests will let you know if you accidentally bork something.

File: definitions.asm (hunk 1 above)
Comment (patricksurry) [reply definitions.asm:53]: Added #62

File: taliforth.asm (hunk 2 above)
Comment (patricksurry) [reply taliforth.asm:230]: oh, interesting.  it does seem at some point that collapsing out `UP` and moving to a "single user" implementation would offer some space/time/complexity saving.   Currently `UP` never changes right?

**Comment by patricksurry** [comment]
OK, this is a working state (but still a little doc cleanup etc).  The ROM is slightly shorter, and unnested loops are decently faster (due to avoiding push/pull of loop control in that case).  `i` is faster.  But nested loops are only a little better (probably would be a slight win overall to remove the unnested loop specialization which would make those slower but avoid the check on each repeat of a nested one).

Just for fun I'm going to try something a bit different which I think will a fair bit quicker and simpler.

File: taliforth.asm (hunk 2 above)
Comment (SamCoVT) [reply taliforth.asm:230]: That is correct - currently UP never changes once set by `quit`

**Comment by patricksurry** [comment]
see #53

**Pull request closed without merging**